### PR TITLE
Circuit cutting: Add expansion to cut_circuit transform

### PR DESCRIPTION
**Pull request by trbromley**
**Context:**

Adds an expansion function to `cut_circuit`, allowing support for circuits that contain `WireCut` operations nested in operations or sub-tapes.

### CODE REVIEW

**Review by anthayes92 — APPROVED** [review]
Thanks @trbromley, a useful addition! Just one question about how the function is defined. 

Well tested, approved :heavy_check_mark:

**Review by josh146 — APPROVED** [review]
Looks good to me @trbromley! Only comment is on the recursion, but not a blocker to merging

**Review by DSGuala — APPROVED** [review]
I tested this with `qml.TTN`.

Trying to follow the Usage Details in [the qcut docs](https://pennylane.readthedocs.io/en/latest/code/api/pennylane.cut_circuit.html) to implement manually, I had to make sure to add an expansion line:

```
n_wires=4
n_wires_device = 2
n_block_wires = 2
n_blocks = 2**(np.log2(n_wires/n_block_wires)+1)-1
n_SEL_layers = 2
n_SEL_wires = n_block_wires

shape = qml.StronglyEntanglingLayers.shape(n_layers=n_SEL_layers, n_wires=n_SEL_wires)
template_weights = [np.random.random(size=shape)] * int(n_blocks)
dev = qml.device('default.qubit',wires=n_wires)

def block(weights, wires):
    qml.StronglyEntanglingLayers(weights,wires)
    qml.WireCut(wires=wires)

with qml.tape.QuantumTape() as tape:
    qml.TTN(
        wires=range(n_wires),
        n_block_wires=n_block_wires,
        block=block,
        n_params_block=3,
        template_weights=template_weights,
    )
    qml.expval(qml.PauliZ(wires=n_wires-1))

###
tape = tape.expand() #this is the expansion line
###
graph = qml.transforms.qcut.tape_to_graph(tape)
qml.transforms.qcut.replace_wire_cut_nodes(graph)
fragments, communication_graph = qml.transforms.qcut.fragment_graph(graph)
fragment_tapes = [qml.transforms.qcut.graph_to_tape(f) for f in fragments]


fragment_tapes = [qml.transforms.qcut.remap_tape_wires(t, dev.wires) for t in fragment_tapes]

expanded = [qml.transforms.qcut.expand_fragment_tape(t) for t in fragment_tapes]

configurations = []
prepare_nodes = []
measure_nodes = []
for tapes, p, m in expanded:
    configurations.append(tapes)
    prepare_nodes.append(p)
    measure_nodes.append(m)

tapes = tuple(tape for c in configurations for tape in c)

results = qml.execute(tapes, dev, gradient_fn=None)

qml.transforms.qcut.qcut_processing_fn(results,communication_graph,prepare_nodes,measure_nodes,use_opt_einsum=True)
```

File: pennylane/transforms/qcut.py [hunk 1]
```diff
@@ -1216,6 +1217,33 @@ def qnode_execution_wrapper(self, qnode, targs, tkwargs):
...
+    if max_depth > 0:
+        return cut_circuit.expand_fn(tape.expand(), max_depth=max_depth - 1)
+
+    raise ValueError(
+        "No WireCut operations found in the circuit. Consider increasing the max_depth value if "
+        "operations or nested tapes contain WireCut operations."
+    )
+
+
+cut_circuit.expand_fn = _qcut_expand_fn
```
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/transforms/qcut.py:1244]: Yes good question. It was originally
```python
return cut_circuit.expand_fn(tape.expand(), max_depth=max_depth - 1)
```
on line 1236, and this works too. It just makes it hard to track with spies in the test, we'd need one spy for `cut_circuit.expand_fn` (first call) and another for `_qcut_expand_fn` (subsequent calls).

**Comment by trbromley** [comment]
Thanks @DSGuala! What is the issue you are seeing with the `tape = tape.expand()` line commented out? It's working ok for me with and without that line.

**Comment by DSGuala** [comment]
> Thanks @DSGuala! What is the issue you are seeing with the `tape = tape.expand()` line commented out? It's working ok for me with and without that line.

@trbromley Sorry, I shared the wrong code. This one should give an error when we comment out that expansion line. The reason is that the tape is not getting cut into fragments.

```
n_wires=4
n_wires_device = 2
n_block_wires = 2
n_blocks = 2**(np.log2(n_wires/n_block_wires)+1)-1
n_SEL_layers = 2
n_SEL_wires = n_block_wires

np.random.seed(1)
shape = qml.StronglyEntanglingLayers.shape(n_layers=n_SEL_layers, n_wires=n_SEL_wires)
template_weights = [np.random.random(size=shape)] * int(n_blocks)
dev = qml.device('default.qubit',wires=n_wires_device)

def block(weights, wires):
    qml.StronglyEntanglingLayers(weights,wires)
    qml.WireCut(wires=wires)

with qml.tape.QuantumTape() as tape:
    qml.TTN(
        wires=range(n_wires),
        n_block_wires=n_block_wires,
        block=block,
        n_params_block=3,
        template_weights=template_weights,
    )
    qml.expval(qml.PauliZ(wires=n_wires-1))

###
#tape = tape.expand() #this is the expansion line
###
graph = qml.transforms.qcut.tape_to_graph(tape)
qml.transforms.qcut.replace_wire_cut_nodes(graph)
fragments, communication_graph = qml.transforms.qcut.fragment_graph(graph)
fragment_tapes = [qml.transforms.qcut.graph_to_tape(f) for f in fragments]


fragment_tapes = [qml.transforms.qcut.remap_tape_wires(t, dev.wires) for t in fragment_tapes]

expanded = [qml.transforms.qcut.expand_fragment_tape(t) for t in fragment_tapes]

configurations = []
prepare_nodes = []
measure_nodes = []
for tapes, p, m in expanded:
    configurations.append(tapes)
    prepare_nodes.append(p)
    measure_nodes.append(m)

tapes = tuple(tape for c in configurations for tape in c)
results = qml.execute(tapes, dev, gradient_fn=None)

qml.transforms.qcut.qcut_processing_fn(results,communication_graph,prepare_nodes,measure_nodes,use_opt_einsum=True)
```

**Pull request merged**